File: src/main/java/play/pluv/config/LogFilter.java
package play.pluv.config;

import jakarta.servlet.Filter;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.ServletRequest;
import jakarta.servlet.ServletResponse;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.io.UnsupportedEncodingException;
import lombok.extern.slf4j.Slf4j;
import org.slf4j.MDC;
import org.springframework.web.util.ContentCachingRequestWrapper;
import org.springframework.web.util.ContentCachingResponseWrapper;

@Slf4j
public class LogFilter implements Filter {

  @Override
  public void doFilter(
      final ServletRequest request, final ServletResponse response, final FilterChain filterChain
  ) throws IOException, ServletException {
    final ContentCachingRequestWrapper httpServletRequest
        = new ContentCachingRequestWrapper((HttpServletRequest) request);
    final ContentCachingResponseWrapper httpServletResponse
        = new ContentCachingResponseWrapper((HttpServletResponse) response);

    filterChain.doFilter(httpServletRequest, httpServletResponse);
    loggingRequest(httpServletRequest);
    loggingResponse(httpServletResponse);
  }

  private static void loggingRequest(final ContentCachingRequestWrapper httpServletRequest)
      throws UnsupportedEncodingException {
    final String uri = httpServletRequest.getRequestURI();

    //request 내용 확인
    final String reqContent = new String(
        httpServletRequest.getContentAsByteArray(),
        httpServletRequest.getCharacterEncoding()
    );
    MDC.put("uri", uri);
    MDC.put("contentBody", reqContent);
    log.info("request logging");
    MDC.clear();
  }

  private static void loggingResponse(
      final ContentCachingResponseWrapper httpServletResponse
  ) throws IOException {
    int httpStatus = httpServletResponse.getStatus();
    final String resContent = new String(httpServletResponse.getContentAsByteArray());

    //주의 : response를 클라이언트에서 볼 수 있도록 하려면 response를 복사해야 한다. response를 콘솔에 보여주면 내용이 사라진다.
    httpServletResponse.copyBodyToResponse();

    MDC.put("httpStatus", String.valueOf(httpStatus));
    MDC.put("contentBody", resContent);
    log.info("response logging");
    MDC.clear();
  }
}

File: src/main/kotlin/play/pluv/config/LogFilter.kt
package play.pluv.config

import jakarta.servlet.*
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.slf4j.LoggerFactory
import org.slf4j.MDC
import org.springframework.web.util.ContentCachingRequestWrapper
import org.springframework.web.util.ContentCachingResponseWrapper
import java.io.IOException
import java.io.UnsupportedEncodingException

class LogFilter : Filter {

    @Throws(IOException::class, ServletException::class)
    override fun doFilter(
        request: ServletRequest, response: ServletResponse, filterChain: FilterChain
    ) {
        val httpServletRequest = ContentCachingRequestWrapper(request as HttpServletRequest)
        val httpServletResponse = ContentCachingResponseWrapper(response as HttpServletResponse)

        filterChain.doFilter(httpServletRequest, httpServletResponse)

        loggingRequest(httpServletRequest)
        loggingResponse(httpServletResponse)
    }

    companion object {

        private val log = LoggerFactory.getLogger(LogFilter::class.java)

        @Throws(UnsupportedEncodingException::class)
        private fun loggingRequest(httpServletRequest: ContentCachingRequestWrapper) {
            val uri = httpServletRequest.requestURI

            //request 내용 확인
            val reqContent = String(
                bytes = httpServletRequest.contentAsByteArray,
                charset = charset(charsetName = httpServletRequest.characterEncoding)
            )

            try {
                MDC.put("uri", uri)
                MDC.put("contentBody", reqContent)
                log.info("request logging")
            } finally {
                MDC.clear()
            }
        }

        @Throws(IOException::class)
        private fun loggingResponse(
            httpServletResponse: ContentCachingResponseWrapper
        ) {
            val httpStatus = httpServletResponse.status
            val resContent = String(httpServletResponse.contentAsByteArray)

            //주의 : response를 클라이언트에서 볼 수 있도록 하려면 response를 복사해야 한다. response를 콘솔에 보여주면 내용이 사라진다.
            httpServletResponse.copyBodyToResponse()

            try {
                MDC.put("httpStatus", httpStatus.toString())
                MDC.put("contentBody", resContent)
                log.info("response logging")
            } finally {
                MDC.clear()
            }
        }
    }
}
